<?xml version="1.0"?>
<launch>
    <arg name='ns' default='mob_rob'/>
    <arg name="robot_x" default="0" doc="mobile robot start x"/>
    <arg name="robot_y" default="0" doc="mobile robot start y"/>
    
    <group ns="$(arg ns)">
        <node name= "map_server" pkg="map_server" type="map_server" args="$(find task_planner)/maps/willowgarage_refined.yaml"/>
        <param name="tf_prefix" value="$(arg ns)" />
        
        <param name="robot_description" command="cat $(find mobile_robot)/urdf/mob_rob.urdf"/>

        <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-file $(find mobile_robot)/urdf/mob_rob.urdf -urdf -model mob_rob -x $(arg robot_x) -y $(arg robot_y)" />

        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
        <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher"/>

        <!-- Start depthimage to laserscan nodelet. ROS Noetic paket: ros-noetic-depthimage-to-laserscan -->
        <node pkg="nodelet" type="nodelet" name="laserscan_nodelet_manager" args="manager"/>

        <node pkg="nodelet" type="nodelet" name="depthimage_to_laserscan" args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet laserscan_nodelet_manager">
            <param name="scan_height" value="10"/>
            <param name="output_frame_id" value="laser_link"/>
            <param name="range_min" value="0.45"/>
            <remap from="image" to="/camera/depth/image_raw"/>
            <remap from="/scan" to="scan"/>
        </node>

        <!-- Start move_base-->
        <node name="move_base" pkg="move_base" type="move_base" respawn="false" output="screen">
            <rosparam file="$(find mobile_robot)/config/costmap_common_params.yaml" command="load" ns="global_costmap"/>
            <rosparam file="$(find mobile_robot)/config/costmap_common_params.yaml" command="load" ns="local_costmap"/>
            <rosparam file="$(find mobile_robot)/config/local_costmap_params.yaml" command="load"/>
            <rosparam file="$(find mobile_robot)/config/global_costmap_params.yaml" command="load"/>
            <rosparam file="$(find mobile_robot)/config/base_local_planner_params.yaml" command="load"/>
        </node>
        
        <node pkg="amcl" type="amcl" name="amcl" output="screen">
            <param name="odom_model_type" value="diff"/>
            <param name="transform_tolerance" value="0.2" />
            <param name="gui_publish_rate" value="10.0"/>
            <param name="laser_max_beams" value="30"/>
            <param name="min_particles" value="1000"/>
            <param name="max_particles" value="10000"/>
            <param name="kld_err" value="0.05"/>
            <param name="kld_z" value="0.99"/>
            <param name="odom_alpha1" value="0.2"/>
            <param name="odom_alpha2" value="0.2"/>
            <param name="odom_alpha3" value="0.8"/>
            <param name="odom_alpha4" value="0.2"/>
            <param name="odom_alpha5" value="0.1"/>
            <param name="laser_z_hit" value="0.75"/>
            <param name="laser_z_short" value="0.05"/>
            <param name="laser_z_max" value="0.05"/>
            <param name="laser_z_rand" value="0.5"/>
            <param name="laser_sigma_hit" value="0.2"/>
            <param name="laser_lambda_short" value="0.1"/>
            <param name="laser_model_type" value="likelihood_field"/>
            <param name="laser_likelihood_max_dist" value="2.0"/>
            <param name="update_min_d" value="0.2"/>
            <param name="update_min_a" value="0.3"/>
            <param name="odom_frame_id" value="$(arg ns)/odom"/>
            <param name="resample_interval" value="1"/>
            <param name="transform_tolerance" value="0.1"/>
            <param name="recovery_alpha_slow" value="0.0"/>
            <param name="recovery_alpha_fast" value="0.0"/>
            
            <remap from="map" to="/map"/>
            <param name='base_frame_id' value='$(arg ns)/base_link'/>
        </node>

        <!-- Publish initial robot state-->
        <node name="initial_pose_publisher" pkg="mobile_robot" type="initial_pose_publisher.py">
            <param name="robot_x" value="$(arg robot_x)"/>
            <param name="robot_y" value="$(arg robot_y)"/>
        </node>
    </group>
</launch>